components/calendar/src/chinese_based.rs

```diff
@@ -253,42 +307,83 @@ impl<C: ChineseBased + CalendarArithmetic> ChineseBasedDateInner<C> {
             "Year should be in range of i32"
         );
         let year = year_int.saturate();
-        let new_moon = Self::new_moon_before((date + 1).as_moment());
-        let month_i64 =
-            libm::round((new_moon - first_day_of_year) as f64 / MEAN_SYNODIC_MONTH) as i64 + 1;
-        debug_assert!(
-            ((u8::MIN as i64)..=(u8::MAX as i64)).contains(&month_i64),
-            "Month should be in range of u8! Value {month_i64} failed for RD {date:?}"
-        );
-        let month = month_i64 as u8;
-        let day_i64 = date - new_moon + 1;
-        debug_assert!(
-            ((u8::MIN as i64)..=(u8::MAX as i64)).contains(&month_i64),
-            "Day should be in range of u8! Value {month_i64} failed for RD {date:?}"
-        );
-        let day = day_i64 as u8;
-        let is_leap_year = Self::new_year_is_leap_year(
-            first_day_of_year,
-            Some(next_solstice),
-            Some(next_new_year),
-        );
-        let leap_month = if is_leap_year {
-            // This doesn't need to be checked for `None`, since `get_leap_month_from_new_year`
-            // will always return a number greater than or equal to 1, and less than 14.
-            NonZeroU8::new(Self::get_leap_month_from_new_year(first_day_of_year))
+        let data_option = C::get_compiled_data_for_year(year);
```

Comment:
this needs `new_year_on_or_before_fixed_date` above, which is expensive

I think that's part of why this is expensive; potentially what we need to do is estimate the chinese year (easy), and then check entries in the neighborhood of the cache. Basically add `C::get_compiled_data_for_rd()` which takes an RD, computes the ISO year, does a simple offset to compute the Chinese year that it is most likely to be in, checks the array at that chinese year, and if it missed by one it can look forward or behind.

Reply:
I've approached this by having the `chinese_based_date_from_fixed` function take an optional `Date<Iso>`; it's possible that I could just replace this entirely with something like `chinese_based_date_from_iso`, since I think the only time this function is called (other than in tests) is in `date_from_iso`. Either way, this greatly reduces the bench time (I'll put an update comment below)

Reply:
ah, that works. Still do think we can clean up further but this is nice